### PR TITLE
 [Schemas] In-memory schema service

### LINKED ISSUE #1561: [Schemas] In-memory schema service
Part of https://github.com/ConduitIO/conduit/issues/1560.

We need to implement a service that exposes methods to fetch and register schemas. The service will be used for other internal, connector-related, methods.  "Connector utility service" is just a working name, we can change to whatever we agree might be better.

It also needs to run on a randomly generated port (in the future this port might be made configurable). The port should be exposed through an environment variable so that connectors can use it.

In this issue, we can have an in-memory implementation, so that the Connector SDK work can be unblocked.

**Note**: We should keep in mind that connectors and processors communicate with Conduit differently.

### PR DESCRIPTION
**Pull request by hariso**
### Description

Closes #1561.  Depends on https://github.com/ConduitIO/conduit-connector-protocol/pull/88. 

### Quick checks:

- [ ] I have followed the [Code Guidelines](https://github.com/ConduitIO/conduit/blob/main/docs/code_guidelines.md).
- [x] There is no other [pull request](https://github.com/ConduitIO/conduit/pulls) for the same update/change.
- [x] I have written unit tests.
- [ ] I have made sure that the PR is of reasonable size and can be easily reviewed.

### CODE REVIEW

File: pkg/conduit/runtime.go
Comment (hariso) [inline pkg/conduit/runtime.go:178]: Previously, the stats handler was manually created because message stats were not needed (since we didn't use streams). We started using streams (between Conduit and connectors), so we can now use this stats handler.

File: pkg/connservices/doc.go
Comment (lovromazgon) [inline pkg/connservices/doc.go:18]: I would pull this into `plugin/connectors`, it's only relevant for connector plugins.

File: pkg/plugin/processor/builtin/impl/avro/schemaregistry/fake.go
Comment (lovromazgon) [inline pkg/plugin/processor/builtin/impl/avro/schemaregistry/fake.go:97]: I think this might also be more usfeul in conduit-commons, I expect we'll need it in tests in the connector and processor SDK.

File: pkg/conduit/runtime.go
Comment (lovromazgon) [reply pkg/conduit/runtime.go:178]: We don't really collect metrics from the plugin using the gRPC stats handler. Regardless, I'm not sure why we had a custom stats handler in the first place tbh, even if we don't have streams, it doesn't really hurt if we register those stream stats collectors 👍

File: pkg/connservices/proto_converter.go
Comment (lovromazgon) [inline pkg/connservices/proto_converter.go:26]: I assume you took the `protoConverter` idea from the processors package. I propose to rather move these converters into the connector-protocol and use the same [`fromproto`](https://github.com/ConduitIO/conduit-connector-protocol/tree/main/cplugin/v2/fromproto) and [`toproto`](https://github.com/ConduitIO/conduit-connector-protocol/tree/main/cplugin/v2/toproto) pattern. A nice side-effect of this is that we won't be able to use any types from Conduit, we'll have to use types that are defined in the protocol or in conduit-commons, which will in turn make it possible to reuse them in connectors and processors.

File: pkg/schema/instance.go
Comment (hariso) [reply pkg/schema/instance.go:29]: A standalone connector will get a Protobuf schema message. From it, we can take the bytes, get the Avro schema of it, and make it available. Similar is true for standalone processors.

File: pkg/plugin/processor/builtin/impl/avro/schemaregistry/fake.go
Comment (hariso) [reply pkg/plugin/processor/builtin/impl/avro/schemaregistry/fake.go:97]: Hmm, but the Connector SDK for example should use a schema registry only through Conduit, i.e. it's Conduit's schema API which we can mock and use in the tests.

File: pkg/schema/instance.go
Comment (hariso) [reply pkg/schema/instance.go:29]: Is it going to be used across all of them though? A standalone connector will get a Protobuf schema message. From it, we can take the bytes, get the Avro schema of it, and make it available. Similar is true for standalone processors.

Comment (lovromazgon) [reply pkg/schema/instance.go:29]: The protobuf message will have to be transformed into a Go type like the one we have here, so the connector developer doesn't have to work with proto types. Sure, we _could_ have separate objects in all repos where we need it, but that's why we introduced conduit commons, so we don't duplicate the types and utilities around them.

Comment (hariso) [reply pkg/schema/instance.go:29]: You're right, we'll need a Go type for the schema. But that will be an Go type representing an Avro schema, like [this](https://pkg.go.dev/github.com/hamba/avro/v2#Schema), no?

File: pkg/plugin/processor/builtin/impl/avro/schemaregistry/fake.go
Comment (lovromazgon) [reply pkg/plugin/processor/builtin/impl/avro/schemaregistry/fake.go:97]: Mocking the registry is a valid approach, but the registry is already a pretty dumb thing, when I was working on that `schemaregistry` package I found it much simpler to work with the fake registry (that's why I added it in the first place). Similarly how we use the in-memory DB in Conduit tests, I think it will be easier to write tests using a proper in-memory schema registry.

Although if you prefer mocking, I won't stand in your way 😉

Comment (lovromazgon) [reply pkg/plugin/processor/builtin/impl/avro/schemaregistry/fake.go:97]: It will call the gRPC service, that's right, I'm saying that in tests we might want to forward the calls to this ~fake~ in-memory registry instead of a mock. In other words, instead of a mock we'd have a small adapter type, which implements the gRPC service interface that the connector is using and calls the in-memory registry.

File: pkg/schema/instance.go
Comment (hariso) [reply pkg/schema/instance.go:29]: Gotcha, and I was about to say that a connector developer probably doesn't need a nice Go type besides an Avro schema itself, but then I remember the schema version for example. So we'll definitely need one.

File: pkg/plugin/processor/builtin/impl/avro/schemaregistry/fake.go
Comment (hariso) [reply pkg/plugin/processor/builtin/impl/avro/schemaregistry/fake.go:97]: I think you're making a good, but at least in this moment, I'm not quite convinced it's worth. So maybe we just leave the FakeRegistry, and move it into commons when/if we realize it's easier to test with it? That way, we'll hopefully keep the commons repo a bit cleaner.

File: pkg/plugin/processor/builtin/impl/avro/schemaregistry/fake.go
Comment (lovromazgon) [reply pkg/plugin/processor/builtin/impl/avro/schemaregistry/fake.go:101]: So the idea is to completely rely on Apicurio? I was actually talking about changing this struct so it becomes an actual schema registry. All it's missing is just a way to permanently store the schemas, that's what I mean by giving it access to a [`database.DB`](https://github.com/ConduitIO/conduit/blob/bf18b8c4649f0c418ba5bc93f92cfbdacd2ae09a/pkg/foundation/database/database.go#L31) object and let it store the schemas there.

What worries me is that if we don't have a schema registry in here, then Conduit will have a hard dependency on Apicurio (well, depending on the connectors used in the pipeline) and won't be a completely standalone tool anymore, which is one of the biggest appeals of Conduit. I'm not saying that we shouldn't use an external registry, we should, that's how schemas can be shared across Conduit instances. But I'm trying to prevent the external registry becoming a hard dependency, things should ideally still work if there is none.

Comment (hariso) [reply pkg/plugin/processor/builtin/impl/avro/schemaregistry/fake.go:101]: Leaving the no-op (in memory) implementation around for development purposes is a good idea, we can do that. :+1: 

As for storing the schemas in the Conduit DB, I think that too is a good idea, but we should discuss it as a separate item, so that expectations are correctly, especially given that we plan a Conduit schema registry long term anyway.

File: pkg/schema/service.go
Comment (hariso) [reply pkg/schema/service.go:67]: Yup, however, in the Apicurio Registry, IDs are strings. If the caller doesn't supply one, they are UUIDs (we may or may not use that functionality). Record metadata values are strings too, so I believe that all in all it's better to leave them strings: it gives us more flexibility.

File: pkg/plugin/processor/builtin/impl/avro/schemaregistry/fake.go
Comment (hariso) [reply pkg/plugin/processor/builtin/impl/avro/schemaregistry/fake.go:99]: > Now that I'm thinking about it, maybe we can get away without having a schema.Service, but just having a schema.Registry.

We might be able to do that... But I think we can have a clearer picture once we're done integrating with Apicurio.:)

File: pkg/plugin/processor/builtin/impl/avro/schemaregistry/fake.go
Comment (hariso) [reply pkg/plugin/processor/builtin/impl/avro/schemaregistry/fake.go]: Moved to the ctor method.

File: pkg/schema/instance.go
Comment (hariso) [reply pkg/schema/instance.go]: Added to commons: https://github.com/ConduitIO/conduit-commons/commit/31b922714201dae61542657bf4364f125c33b314.

File: pkg/plugin/processor/builtin/impl/avro/decode_examples_test.go
Comment (hariso) [inline pkg/plugin/processor/builtin/impl/avro/decode_examples_test.go:15]: The fake registry now needs to be part of the build.

File: pkg/schema/fake_registry.go
Comment (lovromazgon) [inline pkg/schema/fake_registry.go:15]: The fake schema registry was previously meant to only run in unit tests, now we include it in Conduit because we use it as the in-memory implementation. This all makes sense. However, previously _integration_ tests were running against a real schema registry, meaning that the same test was ensuring that it works in a real-world use case, while allowing the developer to run fast unit tests without spinning up a schema registry. Notice that [`integration.go`](https://github.com/ConduitIO/conduit/blob/main/pkg/plugin/processor/builtin/impl/avro/schemaregistry/integration.go) and [`fake.go`](https://github.com/ConduitIO/conduit/blob/main/pkg/plugin/processor/builtin/impl/avro/schemaregistry/fake.go) both define the functions `ExampleSchemaRegistryURL` and `TestSchemaRegistryURL`, so depending on the flag `--tags=integration` the test either targets the fake or a real schema registry. This behavior should be retained.

Essentially we need to extract everything between lines 31 and 88 into a separate file that's guarded with the `go:build` tag I'm commenting on, and then have another file like [`integration.go`](https://github.com/ConduitIO/conduit/blob/main/pkg/plugin/processor/builtin/impl/avro/schemaregistry/integration.go) that just returns the URL to the real schema registry.

File: pkg/schema/fake_registry.go
Comment (lovromazgon) [inline pkg/schema/fake_registry.go:97]: I wonder if `FakeRegistry` is a suitable name now that we don't use it solely as a fake in tests but also as a real in-memory registry.

File: pkg/schema/in_memory.go
Comment (lovromazgon) [inline pkg/schema/in_memory.go:47]:
```suggestion
		Name:    created.Subject,
		Version: created.Version,
```

File: pkg/schema/in_memory.go
Comment (lovromazgon) [inline pkg/schema/in_memory.go:67]:
```suggestion
		Name:    created.Subject,
		Version: created.Version,
```

File: pkg/schema/service.go
Comment (lovromazgon) [inline pkg/schema/service.go:22]: Why are we renaming this to `commschema`? Other files don't do that 🤔

Btw, given the many package name clashes between conduit-commons, protocol, and conduit, did we consider a different naming approach? After thinking about this for a while, this is my proposal:
- Rename this package to `schemaregistry`. I know, not ideal, but it removes the naming conflict.
- The conduit-commons package stays as it is, simply `schema`.
- For the protocol I would introduce the naming convention of prefixing packages with `p` as in "protocol", so the package would become `pschema`. Similarly, we would rename `cplugin` to `pconnector`, to indicate it contains the protocol code for the `connector` proto files.

WDYT? If we do this, I can include the `cplugin` -> `pconnector` change in my PRs.

File: pkg/schema/service.go
Comment (lovromazgon) [inline pkg/schema/service.go:30]: Not for this PR, but I'll mention it before I forget. I realized that having a builtin schema registry could create the expectation that the [`avro.decode`](https://conduit.io/docs/processors/builtin/avro.decode) and [`avro.encode`](https://conduit.io/docs/processors/builtin/avro.encode) processors would be able to use the internal schema registry. It would be a bit strange if we had to configure Conduit to point to a schema registry and then again supply the URL to the processors separately. So ideally the processors should use this service to create and get schemas. Not sure if this was included in the design or not.

File: pkg/plugin/processor/builtin/impl/avro/schemaregistry/internal/cache.go
Comment (lovromazgon) [inline pkg/plugin/processor/builtin/impl/avro/schemaregistry/internal/cache.go:36]: This is a bit of a smell if you ask me, the `Rabin` fingeprint is an implementation detail, that's why it was in the `internal` package before. Seems to me that this cache belongs in the `schema` package. I won't dwell on it though, we can refactor it in a PR down the road when we have a better picture what goes where.

File: pkg/schema/fake_registry.go
Comment (hariso) [reply pkg/schema/fake_registry.go:15]: Things are easy to miss with the build tags.:)

File: pkg/schema/fake_registry.go
Comment (hariso) [reply pkg/schema/fake_registry.go:97]: Good point, I'll rename it!

File: pkg/schema/service.go
Comment (hariso) [reply pkg/schema/service.go:30]: I created an issue for this so we can discuss it there: https://github.com/ConduitIO/conduit/issues/1653.

File: pkg/plugin/processor/builtin/impl/avro/schemaregistry/internal/cache.go
Comment (hariso) [reply pkg/plugin/processor/builtin/impl/avro/schemaregistry/internal/cache.go:36]: I'll add a todo for us here, so we don't forget about it before merging the feature branch to `main`.

File: pkg/schema/service.go
Comment (hariso) [reply pkg/schema/service.go:22]: > Why are we renaming this to `commschema`? Other files don't do that 🤔

`cschema` didn't read accurate to me.:) So I went with `commschema` to try indicating it comes from `commons`.

As for the name changes: yeah, it does look like we're bumping into this more frequently. I would suggest we rename it once this and Raul's PR are merged, so we don't derail the current work too much.

As for the names themselves: how about we add a `c` prefix to the `commons` packages?

File: pkg/plugin/processor/builtin/impl/avro/schemaregistry/internal/cache.go
Comment (hariso) [reply pkg/plugin/processor/builtin/impl/avro/schemaregistry/internal/cache.go:36]: FTR, Raul will fix this in his PR.

File: pkg/schema/fake_registry.go
Comment (hariso) [reply pkg/schema/fake_registry.go]: Done!

File: pkg/schema/service.go
Comment (raulb) [reply pkg/schema/service.go:22]: For context, I'm making some pkg changes in https://github.com/ConduitIO/conduit/pull/1630

File: pkg/plugin/processor/builtin/impl/avro/schemaregistry/internal/cache.go
Comment (raulb) [reply pkg/plugin/processor/builtin/impl/avro/schemaregistry/internal/cache.go:36]: Yup https://github.com/ConduitIO/conduit/pull/1630

File: pkg/schema/service.go
Comment (lovromazgon) [reply pkg/schema/service.go:22]: > As for the names themselves: how about we add a `c` prefix to the `commons` packages?

I wouldn't do that for all packages in commons. We do it for _some_ already (`csync`, `cchan`, `ccontext`) where we anticipate conflicts, but I wouldn't make it a rule for all. I'd also keep `schema` named `schema` in commons, as it can be used by processors and connectors as well, so we expose a nice API for them.

Comment (hariso) [reply pkg/schema/service.go:22]: Sounds good to me. I can make the changes once this PR has been merged and once Raul merges his changes too.

File: pkg/conduit/runtime.go
Comment (lovromazgon) [inline pkg/conduit/runtime.go:503]: Hint: `":0"` will choose a random free port. Or if we want to give the user some control over this, we could do the same as go-plugin and let the user define [min and max ports](https://github.com/hashicorp/go-plugin/blob/721087dad1e61cd41a19a30e5ae740dceb55061d/client.go#L188-L191).

File: pkg/schema/service.go
Comment (lovromazgon) [inline pkg/schema/service.go:28]: I'd move this to `errors.go`, we use this in quite a few packages in Conduit, makes it easier to discover them.

**Review by lovromazgon — APPROVED** [review]
A few nitpicks, but approving to unblock this and move on 👍

File: pkg/schema/in_memory.go
Comment (lovromazgon) [inline pkg/schema/in_memory.go:47]: Btw, why do we use `int32` in `schema.Instance`? I anticipate that it will cause a bunch of casts in our code, also given that we'll expose this to the connector and processor this could simply be an `int` to make it easier to work with.

File: pkg/schema/service.go
Comment (lovromazgon) [reply pkg/schema/service.go:22]: Alright, I'll add the rename of `cplugin` -> `pconnector` after I'm done with my current work 👍

File: pkg/schema/in_memory.go
Comment (lovromazgon) [reply pkg/schema/in_memory.go:47]: I just noticed you changed it to `int` in https://github.com/ConduitIO/conduit-commons/pull/67 👍

File: pkg/conduit/runtime.go
Comment (hariso) [reply pkg/conduit/runtime.go:503]: Yup, yup, the port stuff is going to be handled in #1563.

**Pull request merged**
